src/ynETHRedemptionAssetsVault.sol

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: BSD 3-Clause License
+pragma solidity ^0.8.24;
+
+import {Initializable} from "lib/openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol";
+import {AccessControlUpgradeable} from "lib/openzeppelin-contracts-upgradeable/contracts/access/AccessControlUpgradeable.sol";
+import {ReentrancyGuardUpgradeable} from "lib/openzeppelin-contracts-upgradeable/contracts/utils/ReentrancyGuardUpgradeable.sol";
+import {IynETH} from "src/interfaces/IynETH.sol";
+import {IRedemptionAssetsVault} from "src/interfaces/IRedemptionAssetsVault.sol";
+import {ETH_ASSET, YNETH_UNIT} from "src/Constants.sol";
+
+
+contract ynETHRedemptionAssetsVault is IRedemptionAssetsVault, Initializable, AccessControlUpgradeable {
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ERRORS  ------------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    error TransferFailed(uint256 amount, address destination);
+    error ZeroAddress();
+    error InsufficientAssetBalance(address asset, uint256 requestedAmount, uint256 balance);
+    error ContractPaused();
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ROLES  -------------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    bytes32 public constant REDEEMER_ROLE = keccak256("REDEEMER_ROLE");
+    bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+    bytes32 public constant UNPAUSER_ROLE = keccak256("UNPAUSER_ROLE");
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  VARIABLES  ---------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    IynETH public ynETH;
+    bool private _paused;
+
+    // Initializer with Init struct and roles
+    struct Init {
+        address admin;
+        address redeemer;
+        IynETH ynETH;
+    }
+
+    function initialize(Init memory init)
+        external
+        notZeroAddress(init.admin)
+        notZeroAddress(init.redeemer)
+        notZeroAddress(address(init.ynETH))
+        initializer {
+        __AccessControl_init();
+        _grantRole(DEFAULT_ADMIN_ROLE, init.admin);
+        _grantRole(REDEEMER_ROLE, init.redeemer);
+        _grantRole(PAUSER_ROLE, init.admin);
+        _grantRole(UNPAUSER_ROLE, init.admin);
+
+        ynETH = init.ynETH;
+        _paused = false;
+    }
+
+    receive() external payable {
```

Comment:
Limit caller to the `stakingNodeManager` contract?

Reply:
it adds an extra dependency and theres no harm in extra ETH coming in i'd rather skip having the extra dependency